### ISSUE TITLE
Remake PollWatcher

### ISSUE THREAD

**Issue by passcod** [passcod/rsnotify#12]
For those not familiar with the background who want to help:

What was done:

- Support was achieved for both Linux and OS X.
- Windows native support (#4) may be worked on.
- The polling backend requires the ability to get the `mtime` of a file, which cannot currently be done in stable Rust, so it was dropped.

What is needed:

- To get a replacement for `stat.modified()`.
- To implement anew (or adapt from the old code) a polling backend.

**Comment by octplane** [comment]
Hello,

Two ideas for that:

- Finish integrating fsevent in the package (I've started working on this tonight and this is going more or less ok). Then deprecate `poll` until we have a working solution with stable rust. Most of our users, I reckon, are using OSX or an enabled inotify linux, so leaving out the poll solution could be acceptable (or not ?)

- Use ffi to connect to `fstat` api from `rsnotify` directly and provide a working solution that allows `poll` to work correctly in stable rust.

Without poll, the lib seems to compile pretty much ok (cf https://github.com/octplane/rsnotify/tree/fsevent_integration ) and there doesn't seem to be any outdated package.

**Comment by passcod** [comment]
I think most of the flack could come from Windows users who'll be left out, but that's better than nothing working at all.

**Comment by passcod** [comment]
I'm going to rip out Poll straight up, though. It will make this library not much more than an inotify wrapper until fsevent lands, but that's how it was for a while before, too. Then the work will be to rebuild a polling implementation or two. I'm thinking of applying [some actual CS theory to the problem](http://cstheory.stackexchange.com/questions/10205/efficient-diff-algorithm-for-trees-and-levenshtein-distance), when I have time.

**Comment by passcod** [comment]
I've ripped out Poll and released a 2.0.0-pre1 with a NullWatcher that… does nothing. That should at least help things along, something that compiles but does nothing on unsupported platforms is still better than something that just fails.

**Comment by andelf** [comment]
@octplane I copied your fsevent code and made some changes.  Now its works. (``unwatch`` left unimplemented) =>  https://github.com/andelf/rsnotify 
It was cloned before ``NullWatcher`` introduced，so “4 commits ahead, 1 commit behind passcod:master”.

**Comment by octplane** [comment]
@andelf indeed! I think the merge is not too difficult to perform. The unwatch/drop might be more complicated. I had to fight against threads to make things work correctly. I will not be available until tomorrow morning CET, so go ahead if you have some cycles.

**Comment by maurizi** [comment]
I have been using the poll backend on Windows, so I'll be sad to see it go.
That said, I might try and whip up an initial implementation of using `winapi.rs` to do real file watching and fix #4.

**Comment by retep998** [comment]
For getting the file modified time on Windows call `GetFileAttributesExW` if you have a `Path` or `GetFileInformationByHandle` if you have a `HANDLE` or `FindNextFile` if you're iterating over directory entries, at which point you'll get some file information including the last modified time as a `FILETIME` which is "the number of 100-nanosecond intervals since January 1, 1601 (UTC)."

**Comment by maurizi** [comment]
@retep998 I was looking at using [`ReadDirectoryChangesW`](https://retep998.github.io/doc/kernel32/fn.ReadDirectoryChangesW.html) to just directly get file modification events, so I won't really need to get file modification time.

That said, I need to get a directory handle to use `ReadDirectoryChangesW`, which I think means I need to call [`CreateFileW`](https://retep998.github.io/doc/kernel32/fn.CreateFileW.html).  Do you know a simple way to go from a `&Path` to the `LPCWSTR` `CreateFileW` wants?

**Comment by retep998** [comment]
```Rust
let x: Vec<_> = somepath.as_os_str().encode_wide().chain(Some(0)).collect();
CreateFileW(&x, ...);
```

**Comment by blaenk** [comment]
Niiice thanks for helping us out @retep998 :)

Pretty sure that should be a `Some(0).into_iter()` @maurizi

**Comment by blaenk** [comment]
Nevermind! You don't need the `into_iter()` because `chain` takes an `IntoIterator` :D

**Comment by passcod** [comment]
Just [published v2.0.0](), breaking change is the removal of PollWatcher. Keeping this issue to track progress on re-crafting a polling implementation.

**Comment by BurntSushi** [comment]
I think once https://github.com/rust-lang/rust/pull/25844 lands, you should be able to use crate [`filetime`](https://github.com/alexcrichton/filetime) to resolve your last modified issue.

**Comment by passcod** [comment]
Yes indeed. Ok, so putting this on hold until Rust 1.1 lands. Just a few more weeks!

**Comment by blaenk** [comment]
I think @alexcrichton managed to make this work on stable, in part thanks to his `filetime` package. See [this PR](https://github.com/rust-lang/cargo/pull/1712/files)

**Comment by blaenk** [comment]
Nevermind, it's once 1.1 is released, like we already knew. At least we know what the solution will look like :)

**Comment by blaenk** [comment]
So 1.1 is out, we can move forward on this.

**Comment by blaenk** [comment]
Heads up I'm working on this. I got it working already, but gonna go over it one more time to make sure everything's correct.

**Comment by octplane** [comment]
That's great !
